usdValidation: usdUtils: MissingReferenceValidator

beersandrew opened the pull request:
### Description of Change(s)
- created implementation, test, as well as associated entries in plugInfo.json & validatorTokens.h

### Link to proposal ([if applicable](https://openusd.org/release/contributing_to_usd.html#step-1-get-consensus-for-major-changes))

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3/views/1?pane=issue&itemId=65424971

Note: This PR is a replacement for https://github.com/PixarAnimationStudios/OpenUSD/pull/3403

### Checklist

[X] I have created this PR based on the dev branch

[X] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

[X] I have added unit tests that exercise this functionality (Reference: 
[testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

[X] I have verified that all unit tests pass with the proposed changes

[X] I have submitted a signed Contributor License Agreement (Reference: 
[Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

beersandrew committed:
usdValidation: usdUtils: MissingReferenceValidator

created implementation, test, as well as associated entries in plugInfo.json & validatorTokens.h

beersandrew commented on pxr/usdValidation/usdUtilsValidators/validators.cpp:
```diff
@@ -97,9 +97,44 @@ _PackageEncapsulationValidator(const UsdStagePtr &usdStage)
+            ("Found unresolvable external dependency "
+                "'%s'."), unresolvedPath.c_str())
+        );
+    }
```
@tallytalwar One note here, it looks like complianceChecker.py is doing another check [here](https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/complianceChecker.py#L181) which gets some diagnostic information, but it seems like this process of collecting diagnostic information should be started before opening the stage ([here](https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/complianceChecker.py#L1086)). I tried doing this inside the validator with stage.Reload() but was unable to populate the diagnostics this way.

In my specific test, the errors were similar so I wanted your opinion on if it's possible to do something similar here, or if it's necessary to move the diagnostic part over in this case.

tallytalwar replied:
I think we might not need an explicit validator for this now! https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usdValidation/usdValidation/coreValidators.cpp#L16-L29 should provide coverage for this? Do you want to try running it through the CoreValidator and see if it gives you the appropriate composition error wrapped by a CoreValidation error?

beersandrew replied:
Yep, it does! awesome. I'll close this one out then. Thanks

jesschimein commented:
Filed as internal issue #USD-10493

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/plugInfo.json:
```diff
@@ -3,6 +3,9 @@
             "Info": {
                 "Validators": {
+                    "MissingReferenceValidator": {
+                        "doc": "The composed USD stage should not contain any unresolvable asset dependencies (in every possible variation of the asset), when using the default asset resolver."
```
Lets remove the "every possible variation of the asset" thats part of the usdchecker to iterate through various variants. Validator should just be checking a stage/variant.

tallytalwar replied:
Lets leave it as is, since UsdUtilsComputeAllDependencies operates at Sdf level and will give unresolved references across variants.

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/validatorTokens.h:
```diff
@@ -16,18 +16,21 @@
-#define USD_UTILS_VALIDATOR_KEYWORD_TOKENS                                     \
-    (UsdUtilsValidators)                                                       \
+#define USD_UTILS_VALIDATOR_KEYWORD_TOKENS                                    \
+    (UsdUtilsValidators)                                                      \
```
Not sure if UsdUtilsValidators should be the name of it! (I also want to move this pxr/usdValidation/usdUtilsValidators to be called something else). :/

beersandrew replied:
From my perspective, the naming started from "which folder in the usd code base does this validator depend on?". Now that the validation code is moved out of those folders, this feels like a broader problem of defining validation domains, which may be fine for the others already (usdGeom, usdShade... etc). 

Maybe this validator could go into `pxr/usd/usdValidation`? I'm not sure if it's important for that project to not have external dependencies on usdUtils or not.

tallytalwar replied:
Slightly correction the suffix here is the name of the plugin (the directory name happens to be matching the plugin name :D).

I do plan to move the usdUtilsValidators plugin sometime later so lets leave it as-is for now.

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -113,11 +115,54 @@ TestPackageEncapsulationValidator()
+    const UsdStageRefPtr& stage = UsdStage::CreateInMemory();
+
+    const UsdGeomXform xform = UsdGeomXform::Define(stage, SdfPath("/Xform"));
+    const SdfReference badReference("doesNotExist.usd");
```
Can you also try adding a non-layer asset which is unresolved / missing and see if the validator reports that?

tallytalwar replied:
Might also make sense for the test to include an unresolved layer in a variant which isn't selected in the stage (this is when the composition validator will not report this reference layer in a not-selected variant).

beersandrew replied:
I think i've added the non-layer asset test, but take a look and make sure it's what you wanted. I'll have the variant test added tomorrow

beersandrew commented:
Closing since this work is already complete, see Varun's comment above

tallytalwar reviewed (commented):
A few notes, but otherwise lgtm.

beersandrew committed:
feat: adding an asset reference test

jesschimein commented:
/AzurePipelines run

beersandrew committed:
feat: added unused variant test

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -113,11 +126,102 @@ TestPackageEncapsulationValidator()
+
+    const UsdStageRefPtr& stage = UsdStage::CreateInMemory();
+
+    const UsdGeomXform xform = UsdGeomXform::Define(stage, SdfPath("/Xform"));
```
Though this is fine, but you can also create a prim like:
```
prim = stage.DefinePrim("/Prim", "Xform")
```
which won't require you to include usdGeom/xform.h

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -113,11 +126,102 @@ TestPackageEncapsulationValidator()
+                expectedIdentifier);
+    }
+
+    xform.GetPrim().GetReferences().RemoveReference(badLayerReference);
+    const SdfReference badAssetReference("doesNotExist.jpg");
+    xform.GetPrim().GetReferences().AddReference(badAssetReference);
```
Ahh don't think its legal to have a non-usd as a reference like this.

An asset would be something like this: https://github.com/PixarAnimationStudios/OpenUSD/blob/9b0c13b2efa6233c8a4a4af411833628c5435bde/extras/usd/tutorials/simpleShading/simpleShading.usda#L48

So what you can do is:
```python
>>> from pixar import Sdf, Usd, UsdShade
>>> stage = Usd.Stage.CreateInMemory()
>>> shader = UsdShade.Shader.Define(stage, "/Shader")
>>> assetInput = shader.CreateInput("notFoundAsset", Sdf.ValueTypeNames.Asset)
>>> assetInput.Set("NotFoundTexture.png")
True
>>> stage.ExportToString()
'#usda 1.0\n(\n    doc = """Generated from Composed Stage of root layer \n"""\n)\n\ndef Shader "Shader"\n{\n    asset inputs:notFoundAsset = @NotFoundTexture.png@\n}\n\n'
>>>
```

beersandrew replied:
Thanks! This has been updated

tallytalwar commented on pxr/usdValidation/usdUtilsValidators/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -113,11 +126,102 @@ TestPackageEncapsulationValidator()
+    xform.GetPrim().GetReferences().RemoveReference(badAssetReference);
+    // Set an unresolved dependency on an unused variant
+
+    UsdVariantSets variantSets = xform.GetPrim().GetVariantSets();
+    UsdVariantSet colorsVariantSet = variantSets.AddVariantSet("myVariant");
+
+    UsdGeomXform selectedXform = UsdGeomXform::Define(stage, xform.GetPath().AppendChild(TfToken("selected")));
+    UsdGeomXform notSelectedXform = UsdGeomXform::Define(stage, xform.GetPath().AppendChild(TfToken("notSelected")));
```
Don't think this is correct.

You can refer the authoring variant tutorials here: https://openusd.org/docs/Authoring-Variants.html

to see how you can add a variantSet with 2 variants on a prim, with one having a  good reference and other a bad reference. Have the one which has a good reference be the default set on the prim.

Does that make sense?

beersandrew replied:
Yep, thanks for the help here. Should be good now

beersandrew committed:
fix: address PR comments

- remove unnecessary usage of usdGeom
- corrected how the missing asset test was written
- corrected how the variant test was written
